packages/medusa/src/strategies/product-export.ts

```diff
@@ -0,0 +1,451 @@
+          .retrieve(batchJobId)
+
+        const { listConfig, filterableFields } = transformQuery({ defaultRelations: relations_ ,isList: true})(req, {} as any, (err) => throw err)
+        batchJob.context = {
+          ...(batchJob.context ?? {}),
+          listConfig,
+          filterableFields
+        }
+
+        const batchJobRepo = transactionManager.getCustomRepository(this.batchJobRepository_)
+        batchJobRepo.save(batchJob)
+      */
+      return await this.batchJobService_
+        .withTransaction(transactionManager)
+        .ready(batchJobId /* batchJob after the above todo */)
+    })
+  }
+
+  async processJob(batchJobId: string): Promise<BatchJob> {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const batchJob = await this.batchJobService_
+        .withTransaction(transactionManager)
+        .retrieve(batchJobId)
+
+      const {
+        writeStream,
+        url: key,
+        promise,
+      } = await this.fileService_.getUploadStreamDescriptor({
+        name: "product-export",
+        ext: "csv",
+      })
+
+      const header = await this.buildHeader()
+      writeStream.write(header)
+
+      const { listConfig, filterableFields } = batchJob.context as Context
+      let {
+        limit,
+        offset = 0 as number,
+        ...countConfig
+      } = listConfig as Record<string, any>
+
+      const productCount: number = await this.productService_
+        .withTransaction(transactionManager)
+        .count(listConfig)
+
+      let advancementCount = 0
+
+      while (advancementCount < productCount) {
+        const [products, count] = await this.productService_
+          .withTransaction(transactionManager)
+          .listAndCount(filterableFields, {
+            relations: this.relations_,
+            ...((countConfig as Record<string, unknown>) ?? {}),
+            skip: offset,
+            take: this.BATCH_SIZE,
+            order: { created_at: "DESC" },
+          })
+
+        advancementCount += count
+        offset += count
+
+        products.forEach((product) => {
+          writeStream.write(this.buildLine(product))
+        })
+
+        batchJob.context = {
+          ...batchJob.context,
+          offset,
+          count: productCount,
+          progress: advancementCount / productCount,
+        }
+        const batch = await this.batchJobService_.update(batchJobId, {
+          context: batchJob.context,
+        })
+
+        if (batch.status === BatchJobStatus.CANCELED) {
+          writeStream.end()
+
+          await this.fileService_.delete({ key: key })
+
+          return batchJob
+        }
+      }
+
+      writeStream.end()
+
+      await promise
+
+      batchJob.context.fileKey = key
+
+      const updatedBatchJob = await this.batchJobService_.update(batchJobId, {
+        context: batchJob.context,
+      })
+
+      return await this.batchJobService_.complete(updatedBatchJob)
+    }, "READ UNCOMMITTED")
```

Comment:
suggestion: Add an error handler to the atomic phase which sets the status to `failed` along with whatever error was thrown.

Reply:
I am waiting to know if a batch can fail and be re run from the progress it has done already. Since the file will exists and be populated bit by bit if there is a fail some data will have been written and maybe we can re run it which would change the impl a bit. Wdyt?